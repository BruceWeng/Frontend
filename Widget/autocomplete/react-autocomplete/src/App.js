import './App.css';
import {useEffect, useState} from 'react'
function App() {
  const [users, setUsers] = useState([])
  const [userInput, setUserInput] = useState('')
  async function fetchUsers() {
    try {
      const response = await fetch('https://jsonplaceholder.typicode.com/users')
      const users = await response.json()
      setUsers(users)
    } catch(e) {
      console.error(e)
    }
  }
  useEffect(() => {
    fetchUsers()
  }, [])
  function handleUserInputChange(e) {
    setUserInput(e.target.value)
  }

  return (
    <div className="App">
      <datalist id='users'>
        {users.map((user, i) => (
          <option key={i} name='user'>{user.name}</option>
        ))}
      </datalist>
      <form>
        <input className='UserInput' type='text' list='users' name='user' value={userInput} onChange={handleUserInputChange} autoComplete='off' />
      </form>
    </div>
  );
}

export default App;
