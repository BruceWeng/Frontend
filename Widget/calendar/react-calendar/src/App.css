* {
  box-sizing: border-box;
}

html {
  font-family: sans-serif;
  font-size: 15px;
}

nav {
  display: flex;
  flex-direction: row;
}

button {
  all: unset;
  cursor: pointer;
  width: 5rem;
  height: 2rem;
  border: 0.2rem solid black;
  display: flex;
  justify-content: center;
  align-items: center;
}

.Cell {
  width: 5rem;
  height: 5rem;
  background-color: dodgerblue;
  text-align: center;
 }

 .Cell--highlight {
  width: 5rem;
  height: 5rem;
  background-color: tomato;
  text-align: center; 
 }