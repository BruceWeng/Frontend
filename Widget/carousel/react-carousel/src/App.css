* {
  box-sizing: border-box;
}
html {
  font-family: sans-serif;
  font-size: 15px;
}
ul {
  list-style-type: none;
  padding: 0;
}
button {
  all: unset;
  cursor: pointer;
}
.Carousel__slide-list-container {
  display: flex;
  justify-content: center;
  align-items: center;
}
.Carousel__btn--prev {
  font-size: 3rem;
}
.Carousel__btn--prev::before {
  content: '\2039';
}
.Carousel__btn--next {
  font-size: 3rem;
}
.Carousel__btn--next::before {
  content: '\203A';
}
.Carousel__slide-item {
  padding: 0.5rem;
  margin: 0.5rem;
  width: 20rem;
  height: 20rem;
  display: flex;
  justify-content: center;
  align-items: center;
}
.Carousel__dots {
  display: flex;
  justify-content: center;
}
.Carousel__dot {
  font-size: 1.5rem;
  cursor: pointer;
}
.Carousel__dot::before {
  color: lightgray;
  content: '\25CF';
}
.Carousel__dot--highlight {
  font-size: 1.5rem;
  cursor: pointer;
}
.Carousel__dot--highlight::before {
  content: '\25CF';
}