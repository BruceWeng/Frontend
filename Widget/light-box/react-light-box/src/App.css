html {
  size: 12px;
}
.App {
  height: 100%;
}
.Image__container {
  margin: 2rem;
  border: solid red 5px;
  display: grid;
  grid-template-columns: repeat(5, 1fr);
  grid-template-rows: repeat(2, 1fr);
  grid-gap: 1rem;
  justify-items: center;
  align-items: center;
}
.Image {
  height: 15rem;
  width: 15rem;
  border: solid black 2px;
  cursor: pointer;
}
.Modal__container {
  display: grid;
  position: fixed;
  height: 100%;
  width: calc(100vw);
  z-index: 9999;
  left: 0;
  top: 0;
  background-color: rgba(0, 0, 0, 0.6);
  display: grid;
  grid-template-columns: repeat(1, 1fr);
  justify-items: center;
  align-items: center;
}
.Modal__content {
  display: flex;
  justify-content: space-between;
  height: 15rem;
  width: 15rem;
  padding: .5rem;
}
.Modal__icon--close::before {
  font-size: 1rem;
  content: '\2716';
  color: white;
  cursor: pointer;
}