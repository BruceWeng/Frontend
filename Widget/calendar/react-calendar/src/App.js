import './App.css'
import React, {useReducer} from 'react'

function Calendar() {
  const date = new Date();
  const day_map = ['Sun', 'Mon', 'Tue', 'Wed', 'Thr', 'Fri', 'Sat'];
  const month_map = ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec'];
  const current_year = date.getFullYear();
  const current_month = date.getMonth();
  const current_date = date.getDate();
  const ACTIONS = {
    MONTH_DECREASED: 'month-decreased',
    MONTH_INCREASED: 'month-increased'
  }
  const reducer = (time, action) => {
    switch (action.type) {
      case ACTIONS.MONTH_DECREASED:
        return {
          ...time,
          year: time.year + Math.floor(action.payload.month / 12),
          month: (action.payload.month + 12) % 12 
        }
      case ACTIONS.MONTH_INCREASED:
        return { 
          ...time,
          year: time.year + Math.floor(action.payload.month / 12),
          month: (action.payload.month + 12) % 12 
        }
      default:
        return time;
    }
  }
  let [time, dispatch] = useReducer(reducer, {year: date.getFullYear(), month: date.getMonth()});

  const first_day = new Date(time.year, time.month, 1).getDay();
  const day_count = new Date(time.year, time.month + 1, 0).getDate();

  let table = [[...Array(7).keys()]];
  for (let i = 0; i < 6; i++) table.push(Array(7).fill(undefined));
  for (let day = first_day; day < first_day + day_count; day++) {
    let row = Math.floor(day / 7) + 1;
    let col = day % 7;
    table[row][col] = day - first_day + 1;
  }

  const handlePrevButton = () => {
    dispatch({type: ACTIONS.MONTH_DECREASED, payload: {month: time.month - 1}});
  }

  const handleNextButton = () => {
    dispatch({type: ACTIONS.MONTH_INCREASED, payload: {month: time.month + 1}});
  }

  return (
    <>
      <header><h2>{`${time.year} ${month_map[time.month]}`}</h2></header>
      <nav>
        <button onClick={handlePrevButton}>{'<'}</button>
        <button onClick={handleNextButton}>{'>'}</button>
      </nav>
      <table>
        <thead>
        {table.filter((row, i) => i === 0).map((row, i) => {
            return (
              <tr key={i}>{row.map((col, j) => {
                  return <th key={j} className="Cell">{day_map[col]}</th>
              })}</tr>
            )
          })}
        </thead>
        <tbody>
          {table.filter((row, i) => i !== 0 
              && !(row[0] === undefined && row[row.length-1] === undefined)
            )
            .map((row, i) => {
              return (
                <tr key={i}>{row.map((col, j) => {
                  return <td key={j} 
                    className={col === current_date
                      && time.month === current_month
                      && time.year === current_year 
                      ? "Cell--highlight" 
                      : "Cell"}>{col}</td>
                })}</tr>
              )
            })} 
        </tbody>
      </table>
    </>
  );
} // End of Calendar

function App() {
  return (
    <div className="App">
      <Calendar />
    </div>
  );
}

export default App;
