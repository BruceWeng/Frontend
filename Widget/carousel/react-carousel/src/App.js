import './App.css';
import {useEffect, useReducer} from 'react';

function useCarousel() {
  // initial state: {isLoading: false, images: [], selected_idx: 0}
  const ACTIONS = {
    IMAGES_ISLOADING: 'images-isloading',
    IMAGES_INITIALIZED: 'images-initialized',
    SLIDE_INDEX_INCREASED: 'slide-index-increased',
    SLIDE_INDEX_DECREASED: 'slide-index-decreased',
    SLIDE_INDEX_SELECTED: 'slide-index-selected'
  }
  const reducer = (slide, action) => {
    switch (action.type) {
      case ACTIONS.IMAGES_ISLOADING: {
        return {
          ...slide,
          isLoading: action.payload.isLoading
        }
      }
      case ACTIONS.IMAGES_INITIALIZED: {
        return {
          ...slide,
          images: action.payload.images
        }
      }
      case ACTIONS.SLIDE_INDEX_INCREASED: {
        return {
          ...slide,
          selected_idx: Math.floor((action.payload.selected_idx + 1) % slide.images.length) 
        }
      }
      case ACTIONS.SLIDE_INDEX_DECREASED: {
        return {
          ...slide,
          selected_idx: Math.floor((action.payload.selected_idx - 1 + slide.images.length) % slide.images.length)
        }
      }
      case ACTIONS.SLIDE_INDEX_SELECTED: {
        return {
          ...slide,
          selected_idx: action.payload.selected_idx
        }
      }
      default: {
        return slide;
      }
    }
  }
  const [slide, dispatch] = useReducer(reducer, {isLoading: false, images: [], selected_idx: 0})
  async function fetchImages() {
    try {
      const response = await fetch('https://jsonplaceholder.typicode.com/photos?_start=0&_limit=10')
      dispatch({type: ACTIONS.IMAGES_ISLOADING, payload: {isLoading: false}})
      const json = await response.json()
      dispatch({type: ACTIONS.IMAGES_INITIALIZED, payload: {images: json}})
    } catch(error) {
      dispatch({type: ACTIONS.IMAGES_ISLOADING, payload: {isLoading: false}})
      console.error(`Image fetching failed: ${error}`) 
    }
  }
  useEffect(() => {
    dispatch({type: ACTIONS.IMAGES_ISLOADING, payload:{isLoading: true}})
    fetchImages()
  }, [])
  return [slide, dispatch, ACTIONS]
}

function App() {
  const [slide, dispatch, ACTIONS] = useCarousel()
  const handleBtnPrevClick = () => {
    dispatch({type: ACTIONS.SLIDE_INDEX_DECREASED, payload: {selected_idx: slide.selected_idx}})
  }
  const handleBtnNextClick = () => {
    dispatch({type: ACTIONS.SLIDE_INDEX_INCREASED, payload: {selected_idx: slide.selected_idx}})
  }
  const handleDotClick = (idx) => {
    dispatch({type: ACTIONS.SLIDE_INDEX_SELECTED, payload: {selected_idx: idx}})
  }

  return (
    <div className="App">
      <div className="Carousel__slide-list-container">
        <button className="Carousel__btn--prev" data-testid="btn-prev" onClick={handleBtnPrevClick}></button>
        <ul className="Carousel__slide-list">
          {slide.isLoading 
            ? 'loading...' 
            : slide.images.filter((image, idx) => idx === slide.selected_idx)
            .map((image) => <li key={image.id} 
                                className="Carousel__slide-item" 
                                data-testid="slide-item"
                                style={{backgroundImage: `url("${image.url}")`}}>
                              <div>{image.id}. {image.title}</div>
                            </li>)
          }
        </ul>
        <button className="Carousel__btn--next" data-testid="btn-next"onClick={handleBtnNextClick}></button>
      </div>
      <ul className="Carousel__dots">
        {slide.images.map((image, idx) => <li key={idx}
                                              className={idx === slide.selected_idx 
                                                ? "Carousel__dot--highlight"
                                                : "Carousel__dot"} 
                                              data-testid={`dot_${idx}`}
                                              onClick={() => handleDotClick(idx)}>
                                          </li>)
        }
      </ul>
    </div>
  );
}

export default App
