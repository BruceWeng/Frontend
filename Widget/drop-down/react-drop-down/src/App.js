import {useState} from 'react'
import './App.css';

function App() {
  const [name, setName] = useState('')
  const [email, setEmail] = useState('Eliseo@gardner.biz')
  const [body, setBody] = useState('')

  function handleNameChange(e) {
    const {value} = e.target
    setName(value)
  }
  function handleEmailChange(e) {
    const {value} = e.target
    setEmail(value)
  }
  function handleBodyChange(e) {
    const {value} = e.target
    setBody(value)
  }
  async function handleSubmitClick(e) {
    e.preventDefault()
    alert(JSON.stringify({name, email, body}))
    const result = await postData({name, email, body})
    alert(JSON.stringify(result))
  }
  async function postData(data) {
    const response = await fetch('https://jsonplaceholder.typicode.com/comments', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json'
      },
      body: JSON.stringify(data)
    })
    return await response.json()
  }
  return (
    <div className="App">
      <form onSubmit={handleSubmitClick}>
        <label>
          Name:
          <input data-testid='name-input' type='text' name='name' value={name} onChange={handleNameChange}/>
        </label>
        <label>
          Email:
          <select data-testid='email-select' name='email' value={email} onChange={handleEmailChange}>
            <option data-testid='email-select-option' value='Eliseo@gardner.biz'>Eliseo@gardner.biz</option>
            <option data-testid='email-select-option' value='Jayne_Kuhic@sydney.com'>Jayne_Kuhic@sydney.com</option>
            <option data-testid='email-select-option' value='Nikita@garfield.biz'>Nikita@garfield.biz</option>
          </select>
        </label>
        <label>
          Body:
          <textarea data-testid='body-textarea' name='body' value={body} onChange={handleBodyChange}/>
        </label>
        <input data-testid='submit-input' type='submit' value='Submit'></input>
      </form>
    </div>
  );
}

export default App;
