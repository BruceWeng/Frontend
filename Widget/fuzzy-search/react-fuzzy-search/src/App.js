import './App.css';
import { useEffect, useState } from 'react';

function App() {
  const [users, setUsers] = useState([])
  const [userInput, setUserInput] = useState('')
  const [fuzzyUsers, setFuzzyUsers] = useState([])
  async function fetchUsers() {
    try {
      const response = await fetch('https://jsonplaceholder.typicode.com/users')
      const users = await response.json()
      setUsers(users)
      setFuzzyUsers(users)
    } catch(e) {
      console.error(e)
    }
  }
  useEffect(() => {
    fetchUsers()
  }, [])
  useEffect(() => {
    fuzzyMatchPattern(userInput)
  }, [userInput])
  function handleUserInputChange(e) {
    setUserInput(e.target.value)
  }
  function fuzzyMatchPattern(pattern) {
    let newFuzzyUsers = []
    pattern = pattern.toLowerCase()
    for(let user of users) {
      let start = 0
      let isValid = true
      let target = user.name.toLowerCase()
      for(let char of pattern) {
        target = target.slice(start)
        let next_start = target.indexOf(char)+1
        if(next_start>0 && next_start<target.length+1) start = next_start
        else {
          isValid = false 
          continue
        }
      }
      if(!isValid) continue
      newFuzzyUsers.push(user)
    }
    setFuzzyUsers(newFuzzyUsers)
  }
  return (
    <div className="App">
      <input type='text' name={userInput} onChange={handleUserInputChange}/>
      <ol>
        {
          fuzzyUsers.map((fuzzyUser, i) => (<li key={i}>{fuzzyUser.name}</li>))
        }
      </ol>
    </div>
  );
}

export default App;
