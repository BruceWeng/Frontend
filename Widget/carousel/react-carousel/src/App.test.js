import { render, act } from '@testing-library/react';
import user from '@testing-library/user-event'
import App from './App';

global.fetch = () => Promise.resolve({
  json: () => Promise.resolve([
    {
      "albumId": 1,
      "id": 1,
      "title": "accusamus beatae ad facilis cum similique qui sunt",
      "url": "https://via.placeholder.com/600/92c952",
      "thumbnailUrl": "https://via.placeholder.com/150/92c952"
    },
    {
      "albumId": 1,
      "id": 2,
      "title": "reprehenderit est deserunt velit ipsam",
      "url": "https://via.placeholder.com/600/771796",
      "thumbnailUrl": "https://via.placeholder.com/150/771796"
    },
    {
      "albumId": 1,
      "id": 3,
      "title": "officia porro iure quia iusto qui ipsa ut modi",
      "url": "https://via.placeholder.com/600/24f355",
      "thumbnailUrl": "https://via.placeholder.com/150/24f355"
    }
  ])
})
let app
beforeEach(async () => {
  app = await act(async () => render(<App />))
})
test('renders carousel correctly', () => {
  expect(app).toMatchSnapshot()
});
test('previous button decrease image index', () => {

})
test('next button increase image index', () => {

})
test('slide image matches correct image url', () => {

})
test('test dot current index is highlighted', () => {

})