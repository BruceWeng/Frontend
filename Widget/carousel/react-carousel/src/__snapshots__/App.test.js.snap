// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders carousel correctly 1`] = `undefined`;
